src/mlir/symbol.h

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "ast/ast.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/Types.h"
+#include "mlir/IR/Value.h"
+
+namespace mlir::verona
+{
+  // The symbol table has all declared symbols (with the original node
+  // and the MLIR ounterpart) in a scope. Creating a new scope makes
+  // all future insertions happen at that level, destroying it pops
+  // the scope out of the stack.
+
+  // New scopes are created by creating a new local variable like:
+  //   SymbolScopeT scope(symbolTable);
+  // The destructor pops the scope automatically.
+
+  // We cannot use LLVM's ADT/ScopedHashTable like MLIR's Toy example because
+  // that coped hash table does not allow redefinition, which is a problem when
+  // declaring variables with a type only and then assigning values later.
+  template<class T>
+  class ScopedTable
+  {
+    using MapTy = std::map<std::string, T>;
```

Comment:
Why `map` not `unordered_map`?  Actually, `llvm::SmallDenseMap` probably makes more sense, since these are likely to be very small things.  It's probably worth using a new type for the key as well.  Eventually we're likely to want them to be interned strings, rather than doing string comparison on every lookup.